### ISSUE TITLE
Autoload model classes from `site/models` to make them easy to extend in other models

### ISSUE THREAD

**Issue by fvsch** [k-next/kirby#5968]
## Description

In Kirby 3 I have a custom `DefaultPage` class in `site/models/default.php`, and I am extending that class in other models:

```php
// site/models/default.php
use Kirby\Cms\Page;

class DefaultPage extends Page {
  method1() {}
  method2() {}
}

// site/models/article.php
class ArticlePage extends DefaultPage {
  method3() {}
}
```

To make this work, I had to modify my PHP entrypoint (`public/index.php` in my case) to explicitly require `site/models/default.php`:

```php
// public/index.php
define('__ROOT__', dirname(__DIR__));
require __ROOT__ . '/vendor/autoload.php';
// Default page model is not loaded automatically by Kirby
require __ROOT__ . '/site/models/default.php';
```

I was happy about the Kirby 4 [Default Page Model](https://getkirby.com/releases/4.0/default-page-model) feature announcement, but when trying it out with Kirby 4.0.0-rc.1`, I’m running into an issue:

If I remove the explicit `require` call for `site/models/default.php`, trying to extend that model will fail with:

> Class "DefaultPage" not found

It looks like page classes in `site/models` are not autoloaded. Kirby 4 will load `site/models/default.php` for a page type which doesn't have a more specific model, and that's it.

So if you want to inherit from the `DefaultPage` model, you still have to require it explicitly, whether that's from your root `index.php` or from another model:

```php
// site/models/article.php
require_once __ROOT__ . '/site/models/default.php';

class ArticlePage extends DefaultPage {
  // …
}
```

**Expected behavior**  

It would be great if Kirby could autoload page model classes from `site/models`, so that we developers don’t even have to think about it. ^^

## To reproduce

Here is a reproduction (with a very basic autoloader for `*Page` classes, which must be removed or commented out to see the issue):

https://github.com/fvsch/kirby4-default-model-loading

## Your setup
**Kirby Version**  
4.0.0-rc.1

**Your system (please complete the following information)**
(Skipped, not relevant)

**Comment by fvsch** [comment]
Ultimately it's a question of convenience and of a principle of least surprise. Technically, having to `include_once` or `require_once` another PHP script is not surprising in the PHP world. But it can be surprising when you're in the Kirby world of templates, snippets, configs and models, where everything is autoloaded and auto-included by convention.

My ideal solution would be for Kirby to add a class autoloader for `*Page` classes. But that might not be doable, because the path to the models folder is configurable. That's possibly going to trigger some race condition if the `spl_autoload_register` call is in the `__contruct` method of `Kirby\Cms\App` or later?

A less pure solution would be for Kirby to `include_once` the `site/models/default.php` if it exists, even if the current page's model resolves to a different model class than `DefaultPage`. That would enable extending `DefaultPage` without devs having to add include/require calls. It wouldn't fix things like `class ProductPage extends ArticlePage {}`, but those would be even more rare than extending `DefaultPage`, so maybe that's fine?

**Comment by fvsch** [comment]
It would be nice to improve this, or to improve the documentation around this issue.

Currently the marketing copy on https://getkirby.com/releases/4.0 says:

> Of course, other models can extend the default page model, too.

Which suggests it works out of the box, but it doesn't exactly.

Possible documentation solutions:
- remove that sentence in the marketing page for Kirby 4.0;
- or maybe show how to `require_once 'default.php'; ` from another page model in https://getkirby.com/docs/guide/templates/page-models#default-page-model?

<img width="1178" alt="Screenshot 2023-11-28 at 12 15 54" src="https://github.com/getkirby/kirby/assets/243601/5329f1cd-65ad-4f80-b0fc-863594914586">

**Comment by lukasbestle** [comment]
I agree that autoloading the model classes from `site/models` is a good solution as it will "just work" and be transparent to the user. A possible unintended consequence could be that side effects in model files will also be executed (e.g. if the model file also contains random PHP code outside of the class definition). But I would say that such side effects should be very uncommon, so it's likely not something to worry about.

**Comment by distantnative** [comment]
@fvsch Do you experience the problem still with Kirby 4.3? I couldn't pinpoint where/how this got fixed but I cannot reproduce it currently anymore.

**Comment by fvsch** [comment]
@distantnative I still experience the error with 4.3.

If you tried the minimal reproduction at https://github.com/fvsch/kirby4-default-model-loading and updated `getkirby/cms` to 4.3.0, maybe you got confused by how the reproduction is written. As explained in its README, it contains a workaround, so running `composer run dev` without modifying any code won't show the issue. You have to comment out the workaround to see the original error.

I've updated the minimal reproduction to Kirby 4.3, and modified the code in `public/index.php` to make sure the error shows up by default. See the updated README for details.

**Comment by fvsch** [comment]
Another workaround for people using Composer. It's also possible to autoload page model classes with this config in your `composer.json`:

```json
"autoload": {
    "classmap": ["site/models/"]
},
```

(Docs: https://getcomposer.org/doc/04-schema.md#classmap)

When you run `composer install`, this will find all the PHP files within the `site/models` directory and look at the classes defined in them, and include them in the generated class maps in:

```
vendor/
  composer/
    autoload_classmap.php
    autoload_static.php
```

So if you have a `DefaultPage` class in `site/models/default.php`, it will be included in these classmaps. Other model classes (e.g. an `ArticlePage` class from `site/models/article.php`) will also be included, but this doesn't seem to conflict with the way that Kirby autoloads page model classes in my experience, so that should be fine.

Alternatively, only point to the `DefaultPage` class:

```json
"autoload": {
    "classmap": ["site/models/default.php"]
},
```
